docs/book/src/roadmap.md

```diff
@@ -3,32 +3,37 @@
 This roadmap is a constant work in progress, subject to frequent revision. Dates are approximations.
 
 
-## v0.3.7 (v1alpha3+) ~ June/July 2020
+## v0.4 (v1alpha4) ~ Q1 2021
 
 |Area|Description|Issue/Proposal|
 |--|--|--|
-|Testing|E2E Test plan|[Spreadsheet](https://docs.google.com/spreadsheets/d/1uB3DyacOLctRjbI6ov7mVoRb6PnM4ktTABxBygt5sKI/edit#gid=0)
-|Testing|Enable webhooks in integration tests|[#2788](https://github.com/kubernetes-sigs/cluster-api/issues/2788)|
-|Control Plane|KubeadmControlPlane robustness|[#2753](https://github.com/kubernetes-sigs/cluster-api/issues/2753)|
-|Control Plane|KubeadmControlPlane adoption|[#2214](https://github.com/kubernetes-sigs/cluster-api/issues/2214)|
-|Extensibility|Clusterctl library should support extensible templating|[#2339](https://github.com/kubernetes-sigs/cluster-api/issues/2339)|
-|Cluster Lifecycle|ClusterResourceSet experiment|[#2395](https://github.com/kubernetes-sigs/cluster-api/issues/2395)|
-|Core Improvements|Library to watch remote workload clusters|[#2414](https://github.com/kubernetes-sigs/cluster-api/issues/2414)|
-|API, UX|Support and define conditions on cluster api objects|[#1658](https://github.com/kubernetes-sigs/cluster-api/issues/1658)|
-|Extensibility, Infrastructure|Support spot instances|[#1876](https://github.com/kubernetes-sigs/cluster-api/issues/1876)|
-|Extensibility|Machine pre-deletion hooks|[#1514](https://github.com/kubernetes-sigs/cluster-api/issues/1514)|
-
-## v0.4 (v1alpha4) ~ Q4 2020
-
-|Area|Description|Issue/Proposal|
-|--|--|--|
-|UX, Bootstrap|Machine bootstrap failure detection|[#2554](https://github.com/kubernetes-sigs/cluster-api/issues/2554)|
+|Providers|Move to a single manager watching all namespaces for each provider|[#3042](https://github.com/kubernetes-sigs/cluster-api/issues/3042)
+|Clusterctl|Redefine the scope of clusterctl move|[#3354](https://github.com/kubernetes-sigs/cluster-api/issues/3354)
 |Extensibility|Support pluggable machine load balancers|[#1250](https://github.com/kubernetes-sigs/cluster-api/issues/1250)|
```

Comment:
I'm still not 100% convinced this shouldn't be an infrastructure provider concern (eg. vpshere), not sure what needs to change, if anything, in the core providers to support this

Reply:
My thinking here is that you would have cluster.spec.loadBalancerRef, and this would be what's used to create & manage a load balancer for the control plane going forward. If set, whatever controller is satisfying the ref would lifecycle manage a load balancer, including adding/removing machines as backends. The load balancer provider would be responsible for the control plane endpoint (it would either generate it, or the user would have to supply it, depending on the situation). A load balancer provider could also help with quiesced removal of machines as part of a rolling upgrade of a control plane - we could orchestrate removing a machine from the LB so it doesn't receive any more traffic, then delete the Machine.

While the infra providers currently do all this work, it's all "internal" or specific to each infra provider, and there's no way to mix and match. For example, maybe for whatever reason you want to use an HAproxy/nginx/envoy/whatever load balancer with CAPA/CAPZ/CAPG/etc for your control plane LB. There is no way to do that today, because each infra provider forces you to use what it supplies.

Reply:
+1 to what @ncdc said, additionally, it would also make it easier to add support for new/different load balancer types for individual cloud providers as well, such as NLB support for AWS.

To take it a step further, this also greatly enables additional use cases for on-premises use cases, whether it is OpenStack, vSphere, libvirt, or even bare metal use cases. While the major cloud providers could abstract this away under the infrastructure provider implementation, it would still require a lot of duplication across all of the various infrastructure providers that do not have native load balancer support (or it is optional, limited to certain skus, etc). This really feels like a common abstraction that we should provide.